### PR TITLE
Introduce `QueryTransform` and add `HyDEQueryTransform`

### PR DESCRIPTION
**Pull request by Disiok**
### Summary
* Introduce concept of `QueryProcessor` and simple implementation of HyDE query processor 
* Naive integration with `QueryRunner` where we run query processor once
* Add HyDE example to notebook for demo

### TLDR
```
query_processor = HyDEQueryProcessor(include_original=True)
response = index.query("what did author do after going to RISD", query_processor=query_processor)
``` 

### Future work
- Support generating multiple hypothetical docs
- Consider refactor to support index-specific query processing

### CODE REVIEW

File: gpt_index/indices/query/query_runner.py
Comment (jerryjliu) [inline gpt_index/indices/query/query_runner.py:81]: was going to ask this, but this makes sense. i'm potentially interested in how we can do query processing given contents of the nodes themselves

File: examples/vector_indices/SimpleIndexDemo.ipynb
Comment (jerryjliu) [inline examples/vector_indices/SimpleIndexDemo.ipynb:154]: The notebook organization is a mess atm, but imo I'd create a new folder under `examples` called QueryTransformations which specifically demos HyDE.

File: gpt_index/indices/query/query_processor.py
Comment (jerryjliu) [inline gpt_index/indices/query/query_processor.py:23]: would love to have a How-To Guide + Reference section in the docs. 

reference could go in docs/reference/query.rst

**Review by jerryjliu — COMMENTED** [review]
this looks pretty good! i'll approve after the first PR lands

File: gpt_index/indices/query/query_runner.py
Comment (Disiok) [reply gpt_index/indices/query/query_runner.py]: Yep, going to keep this PR small. Definitely interesting future work to improve query processing with node content (and or metadata).

File: examples/vector_indices/SimpleIndexDemo.ipynb
Comment (Disiok) [reply examples/vector_indices/SimpleIndexDemo.ipynb]: Added a separate notebook with HyDE examples.

File: gpt_index/indices/query/query_processor.py
Comment (Disiok) [reply gpt_index/indices/query/query_processor.py]: Updated reference.

**Comment by Disiok** [comment]
@jerryjliu  See the notebook for 3 examples comparing normal vs. HyDE.

**Review by jerryjliu — APPROVED** [review]

**Pull request merged**
